lib/galaxy/tools/evaluation.py

```diff
@@ -85,7 +85,7 @@ def validate_inputs(input, value, context, **kwargs):
         if get_special:
             special = get_special()
             if special:
-                out_data["output_file"] = special.fda
+                out_data["output_file"] = getattr(special, 'fda', None)
```

Comment:
```suggestion
                out_data["output_file"] = getattr(special, 'fda', fda)
```
`output_file` should be similar to a `HistoryDatasetAssociation` and have a dataset attribute that points to an actual Dataset instance.

Since `get_special` is only called in this one place, could we just return the right thing in get_special ?

Reply:
So, does `get_special` mean this: "return an instance of FakeDatasetAssociation using the appropriate dataset or return None"? And the dataset for the fda would be either provided by JobExportHistoryArchive (inside the `fda` property) or by GenomeIndexToolData (which has a dataset)? If so, maybe we should rename it to `get_fake_dataset_association()`?
(there is [similar logic here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/__init__.py#L2014), but it's only partially the same, and it does not include GenomeIndexToolData)

Reply:
i.e.,
```
def get_special_dataset_association():
    # try getting a JobExportHistoryArchive; return its fda
    # if not, try getting a GenomeIndexToolData; return its fda (which I"ll add as a property to that class)
    # if not, return none
```
And then, in tools/evaluation:
```
...
if special:
    out_data["output_file"] = special
```

Reply:
Does https://github.com/galaxyproject/galaxy/commit/578fb56533b6e2d6c2aecc1f2699d78ca7648db2 makes sense ? Feel free to cherry-pick that commit if you agree.

Reply:
Sorry, got the letters mixed up, https://github.com/galaxyproject/galaxy/commit/7991ee0f3d is better.

Reply:
Yep, and the commit description is very helpful (I think I understand the logic now - thanks!). Will cherry-pick now.

Reply:
Done, and fixed the cheery-picking/authorship (Apparently, when I use the `-n` no commit option, it creates a new commit instead of the original. Obviously.)